File: 0003-longest-substring-without-repeating-characters/solution.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        seen_chars = {}
        max_length = 0
        start = 0

        for end in range(len(s)):
            if s[end] in seen_chars and seen_chars[s[end]] >= start:
                start = seen_chars[s[end]] + 1
            seen_chars[s[end]] = end
            max_length = max(max_length, end - start + 1)

        return max_length

File: 0003-longest-substring-without-repeating-characters/solution.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int left = 0;
        int right = 0; 
        Set<Character> set = new HashSet<>();
        int max = 0;
        while(left < s.length() && right < s.length()){
            if(!set.contains(s.charAt(right))){
                set.add(s.charAt(right));
                right++;
            }else{
                set.remove(s.charAt(left));
                left++;
            }
            max = Math.max(max, right - left);
        }

        return max;
    }
}
